### PR TITLE
add yurtappset v1beta1 proposal

### PR DESCRIPTION
**Pull request by luc99hen**
/kind question
> /kind design
> /sig ai
> /sig iot
> /sig network
> /sig storage

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+### Goals
+
+1. Introduce YurtAppSet v1beta1 to replace YurtAppSet v1alpha1, YurtAppDaemon and YurtAppOverrider
+
+
+### Non-Goals
+
+1. Maintain backward compatibility with YurtAppSet v1alpha1
+
```
Comment (rambohe-ch) [inline docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md:35]: Do not remove YurtAppDaemon and YurtAppOverrider in this proposal

File: docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+## Proposal
+
+### Solution Introduction
+
+Users can use YurtAppSet to distribute the same `WorkloadTemplate` (Deployment/Statefulset) to different nodepools by `NodePoolSelector` or nodepool name (NodePoolSelector). Users can also customize the configuration of workloads in different node pools through `WorkloadTweaks`.
```
Comment (rambohe-ch) [inline docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md:41]: nodepool name (NodePoolSelector) --> nodepool name slices?

File: docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+## Motivation
+
+1. One crd instead of three is more user-friendly to users 
+2. In the previous technical route, the rendering of the workload came from both yurtappset and yurtappoverrider, and the version management of the workload needed to take into account the changes from both parts, which was very complex.
+
+
+### Goals
+
+1. Introduce YurtAppSet v1beta1 to replace YurtAppSet v1alpha1, YurtAppDaemon and YurtAppOverrider
+
```
Comment (rambohe-ch) [inline docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md:30]: Workloads are dynamically managed in response to changes in nodepools, this means when new nodepools are established and meet the NodePoolSelector in YurtAppSet, workloads can be automatically created and assigned to them. Conversely, when existing nodepools no longer fulfill the required NodePoolSelector, associated workloads will be removed accordingly.

File: docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+const (
+	// AppDispatched means all the expected workloads are created successfully.
+	AppSetAppDisPatchced YurtAppSetConditionType = "AppDispatched"
+	// AppDeleted means all the unexpected workloads are deleted successfully.
+	AppSetAppDisDeleted YurtAppSetConditionType = "AppDeleted"
+	// AppUpdated means all expected workload are updated successfully.
+	AppSetAppUpdated YurtAppSetConditionType = "AppUpdated"
+	// PoolFound is added to a YurtAppSet when one of owned workloads has failure during its own reconciling.
+	AppSetPoolFound YurtAppSetConditionType = "PoolFound"
```
Comment (rambohe-ch) [inline docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md:205]: PoolFound --> AppReady? this means all workloads of this yurtappset are ready or not?

File: docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+)
```
Comment (rambohe-ch) [inline docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md:206]: - some other conditions:

AppSetPoolNotFound: no nodepools meets the nodepoolselector of yurtappset
AppSetTemplateInvalid: template in yurtappset is invalid

Comment (luc99hen) [reply docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md:206]: I have different opinions here:
- AppSetPoolNotFound is equivalent to AppSetPoolFound conditionstatus=false.
- it's better to put templateInvalid into the condition reason/message in AppDispatched/AppUpdated condition

File: docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md (hunk 4 above)
Comment (luc99hen) [reply docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md:205]: Yes, we can add AppReady to indicate all workloads of this yurtappset are ready

File: docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md (hunk 3 above)
Comment (luc99hen) [reply docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md:32]: done

File: docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md (hunk 2 above)
Comment (luc99hen) [reply docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md]: done

File: docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md (hunk 5 above)
Comment (rambohe-ch) [reply docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md:212]: @luc99hen I agree with your ideas.

**Comment by rambohe-ch** [comment]
@luc99hen @gnunu @Congrool I want to add conclusions that we had made on the meeting for these two points.

> 24.1.10 Community Discussion Results:
> 
> 1. Before using YurtAppSet, when creating node pools, it is recommended that users plan node pool labels in advance to facilitate subsequent node pool management.  @gnunu

A common label will be added to nodepool in the webhook when new nodepool is created, so end user can use this common label to select all nodepools in the YurtAppSet without managing the labels of nodepool.

> 2. Is it better to introduce a new CRD rather than upgrading YurtAppSet? @Congrool
> 
> * Pros: no need to care about YurtAppSet v1alpha1 dependencies
> * Cons: need to propose a new CRD name

because YurtAppSet v1beta1 can keep part consistency with YurtAppSet v1alpha1, at the same time, YurtAppSet will cover the feature of YurtAppDamon and YurtAppOverrider.  I don't import a new CRD for the YurtAppSet v1beta1.

**Review by rambohe-ch — APPROVED** [review]

**Pull request merged**